### PR TITLE
[WIP] Implement wlr_seat

### PR DESCRIPTION
**Pull request by nyorain**
This implements a first version of wlr_seat.
It currently relies on the (nonexistent) wlr_wl_{touch, keyboard, pointer} so it will not compile. But since there are a few design decisions done here, some feedback would be nice.

Is it acceptable to introduce wlr_wl_{touch, keyboard, pointer} or should those interfaces be merged with wlr_{touch, keyboard, pointer}? Latter would be possible but introduce some design issues.

### CODE REVIEW

**Comment by nyorain** [comment]
Will implement the wlr_seat_touch* functions when everything else is alright.
Sadly i cannot really test this yet since usually wayland applications require a keymap.

**Comment by SirCmpwn** [comment]
A keymap is available via shared.h.

File: include/wlr/types/wlr_seat.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		struct wl_resource *surface, wl_fixed_t surface_x, wl_fixed_t surface_y);
+void wlr_seat_pointer_leave(struct wlr_seat *seat, uint32_t serial, struct wl_resource *surface);
+void wlr_set_pointer_button(struct wlr_seat *seat, uint32_t serial, uint32_t time,
+		uint32_t button, uint8_t state);
+
+void wlr_seat_keyboard_keymap(struct wlr_seat *seat, uint32_t format, int fd, uint32_t size);
+void wlr_seat_keyboard_repeat_info(struct wlr_seat *seat, int32_t rate, int32_t delay);
+void wlr_seat_keyboard_key(struct wlr_seat *seat, uint32_t serial, uint32_t time,
+	uint32_t key, uint8_t state);
+void wlr_seat_keyboard_focus(struct wlr_seat *seat, uint32_t serial, struct wl_resource *surface,
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_seat.h]: Remember the part when I said that wlr_seat *wouldn't* handle focus semantics?

File: include/wlr/types/wlr_seat.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	struct wl_array *keys);
+void wlr_seat_keyboard_modifiers(struct wlr_seat *seat, uint32_t serial, uint32_t mods_depressed,
+		uint32_t mods_latched, uint32_t mods_locked, uint32_t group);
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_seat.h]: Inconsistent indentation on functions split across lines - use two indents consistently.

File: types/wlr_seat.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+static const struct wl_seat_interface seat_implementation = {
+	.get_pointer = seat_get_pointer,
+	.get_keyboard = seat_get_keyboard,
+	.get_touch = seat_get_touch,
+	.release = destroy_resource,
+};
+
+static void seat_destroy(struct wl_resource *resource) {
+	struct wlr_seat *seat = wl_resource_get_user_data(resource);
+	for(unsigned int i = 0; i < seat->resources.length; ++i) {
```
Comment (SirCmpwn) [inline types/wlr_seat.c]: Style: add a space before ( and use size_t

File: include/wlr/types/wlr_seat.h (hunk 1 above)
Comment (nyorain) [inline include/wlr/types/wlr_seat.h]: but how to know then to which resource the events should be sent?

File: include/wlr/types/wlr_seat.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	struct {
+		struct wl_signal cursor_set;
+		struct wl_signal start_drag;
+		struct wl_signal set_selection;
+	} events;
+
+	uint32_t caps;
+	const char *name;
+	list_t resources;
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_seat.h]: Use `list_t*`

Comment (SirCmpwn) [inline include/wlr/types/wlr_seat.h]: Better yet, why not a `wl_list`?

File: include/wlr/types/wlr_seat.h (hunk 1 above)
Comment (SirCmpwn) [inline include/wlr/types/wlr_seat.h]: The caller will tell you which client to send the event to, and you can look up the client in your resource list via `wl_resource_get_client`.

File: types/wlr_seat.c [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+void wlr_seat_destroy(struct wlr_seat *seat) {
+	wl_global_destroy(seat->global);
+	free(seat);
+}
+
+void wlr_seat_set_caps(struct wlr_seat *seat, uint32_t caps) {
+	seat->caps = caps;
+	for(unsigned int i = 0; i < seat->resources.length; ++i) {
```
Comment (SirCmpwn) [inline types/wlr_seat.c]: Same style issues here

File: types/wlr_seat.c [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+	if (!seat->pointer_over_res || !seat->pointer_over_res->pointer) {
+		return;
+	}
+
+	wl_pointer_send_motion(seat->pointer_over_res->pointer, time, surface_x, surface_y);
+}
+
+void wlr_seat_pointer_enter(struct wlr_seat *seat, uint32_t serial,
+		struct wl_resource *surface, wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	// TODO: send pointer_leave event if pointer_over_* is not NULL?
```
Comment (SirCmpwn) [inline types/wlr_seat.c]: Don't track this. It's the caller's responsibility.

Comment (SirCmpwn) [inline types/wlr_seat.c]: This interface should be dumb and should just execute the orders of the caller. It shouldn't try to manage focus *at all*.

File: types/wlr_seat.c [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+	struct wlr_seat_resources *res = wlr_seat_resources_for_client(seat,
+		wl_resource_get_client(surface));
+	seat->pointer_over_surf = surface;
```
Comment (SirCmpwn) [inline types/wlr_seat.c]: Likewise, don't track this.

**Comment by SirCmpwn** [comment]
I pushed a branch with the initial work on a different approach. Want to pick that up?

https://github.com/SirCmpwn/wlroots/tree/wlr_seat

**Comment by nyorain** [comment]
So we leave focus to the user after all? Fine by me, the new interface looks good and cleaner.
I guess everything left to do for wlr_seat is fixing up the TODO's and integrate it with the example compositor?

**Comment by SirCmpwn** [comment]
Aye.

**Comment by nyorain** [comment]
Will finish it then in a new pr (since i don't like overriding all the commits here) later today.

**Pull request closed without merging**